tensorflow/lite/micro/kernels/embedding_lookup.cc

```diff
@@ -0,0 +1,213 @@
+  const int num_rows = value->dims->data[0];
+  if (num_rows == 0) {
+    // Propagate empty tensor if input is empty
+    return kTfLiteOk;
+  }
+  const size_t row_bytes = op_data.num_columns * TfLiteTypeGetSize(value->type);
+
+  int8_t* output_raw = tflite::micro::GetTensorData<int8_t>(output);
+  const int8_t* value_raw = tflite::micro::GetTensorData<int8_t>(value);
+  const int32_t* lookup_data = tflite::micro::GetTensorData<int32_t>(lookup);
+  for (int i = 0; i < lookup->dims->data[0]; i++) {
+    int32_t idx = lookup_data[i];
+    if (idx >= num_rows || idx < 0) {
+      MicroPrintf(
+          "EMBEDDING_LOOKUP: index out of bounds. "
+          "Got %d, and bounds are [0, %d]",
+          idx, num_rows - 1);
+      return kTfLiteError;
+    } else {
+      std::memcpy(output_raw + i * row_bytes, value_raw + idx * row_bytes,
+                  row_bytes);
+    }
+  }
+
+  return kTfLiteOk;
+}
+
+TfLiteStatus EvalHybrid(const OpData& op_data, const TfLiteEvalTensor* lookup,
+                        const TfLiteEvalTensor* value,
+                        TfLiteEvalTensor* output) {
+  const int num_rows = value->dims->data[0];
+  const size_t num_colums = op_data.num_columns;
+
+  float* output_ptr = tflite::micro::GetTensorData<float>(output);
+  const int8_t* value_ptr = tflite::micro::GetTensorData<int8_t>(value);
+  const int32_t* lookup_data = tflite::micro::GetTensorData<int32_t>(lookup);
+
+  for (int i = 0; i < lookup->dims->data[0]; i++) {
+    int32_t idx = lookup_data[i];
+    if (idx >= num_rows || idx < 0) {
+      MicroPrintf(
+          "EMBEDDING_LOOKUP: index out of bounds. "
+          "Got %d, and bounds are [0, %d]",
+          idx, num_rows - 1);
+      return kTfLiteError;
+    } else {
+      // Dequantize embedding values.
+      Dequantize(&value_ptr[idx * num_colums], num_colums, op_data.scale, 0,
+                 &output_ptr[i * num_colums]);
+    }
+  }
+
+  return kTfLiteOk;
+}
+
+TfLiteStatus Eval(TfLiteContext* context, TfLiteNode* node) {
+  const TfLiteEvalTensor* lookup =
+      tflite::micro::GetEvalInput(context, node, kInputTensor_0);
+  const TfLiteEvalTensor* value =
+      tflite::micro::GetEvalInput(context, node, kInputTensor_1);
+  TfLiteEvalTensor* output =
+      tflite::micro::GetEvalOutput(context, node, kOutputTensor);
+
+  OpData& op_data = *static_cast<OpData*>(node->user_data);
+
+  switch (value->type) {
+    case kTfLiteFloat32:
+      return EvalSimple(op_data, lookup, value, output);
+    case kTfLiteInt8:
+      if (output->type == kTfLiteFloat32) {
+        return EvalHybrid(op_data, lookup, value, output);
+      } else {
+        return EvalSimple(op_data, lookup, value, output);
+      }
+    default:
+      MicroPrintf("EMBEDDING_LOOKUP only supports FLOAT32 and INT8, got %s.",
+                  TfLiteTypeGetName(output->type));
+      return kTfLiteError;
+  }
+}
+
+}  // namespace
+
+TfLiteRegistration_V1 Register_EMBEDDING_LOOKUP() {
```

Comment:
this will now need to be TFLMRegistration

Reply:
done